### PR TITLE
planner: fix the issue that cannot find column if using question marker in group-by-clause

### LINKED ISSUE #53872: Param as column alias can't be parsed in group statement
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
create table test(id int, col int);
prepare stmt from "select id, ? as col1 from test where col=? group by id,col1";
set @a=100, @b=100;
execute stmt using @a,@b;
```
### 2. What did you expect to see? (Required)

No error

### 3. What did you see instead (Required)

ERROR 1105 (HY000): Unknown column '100' in 'group statement'

### 4. What is your TiDB version? (Required)

v8.1.0

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #53872

Problem Summary: planner: fix the issue that cannot find column if using question marker in group-by-clause

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/types/parser_driver/value_expr.go [hunk 1]
```diff
@@ -240,6 +240,14 @@ type ParamMarkerExpr struct {
...
 	Order     int
 	InExecute bool
+
+	// For "select ? as c from t group by c", the optimizer replaces the `c` in the by-clause to `group by ?`,
+	// but this conversion conflicts with the original semantic. The original `group by c` means grouping by the column `c`,
+	// while the converted `group by ?` means grouping by the `?-th` column in the select-list, for example, `group by 3` means
+	// grouping the result by the 3rd column.
+	// Use this flag to let the optimizer know whether `group by ?` is converted from this case and if it is, use this
+	// marker as normal value instead of column index in the by-clause.
+	UseAsValueInByClause bool
```
Comment (AilinKid) [inline pkg/types/parser_driver/value_expr.go:250]:
```suggestion
	UseAsValueInGbyByClause bool
```

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

**Review by AilinKid — APPROVED** [review]

**Review by elsa0520 — APPROVED** [review]
LGTM

**Pull request merged**